### PR TITLE
Add a dims field to LinearMixedModel

### PR DESCRIPTION
**Pull request by dmbates**
Dimensions `n` (number of observations) and `p` (rank of the fixed-effects model matrix) are used in every objective function evaluation and require some trivial amount of computation to evaluate.  These are now stored in a `NamedTuple` field called `dims`.

Also, the number of random effects terms (after amalgamation) is needed for every objective function evaluation and requires a non-trivial amount of computation in the current formulation from the `allterms` field.  It is also stored in `dims`.

Using this branch and v1.5.0-rc1 the amount of storage allocated for an objective function evaluation is very small.  For example, with Julia-1.6.0-DEV.360
```julia
julia> mrk17 = MixedModels.dataset(:mrk17_exp1);

julia> contr = merge(
           Dict(nm => HelmertCoding() for nm in (:F,:P,:Q,:lQ,:lT)),
           Dict(nm => MixedModels.Grouping() for nm in (:subj, :item)),
       );

julia> m6form = @formula 1000/rt ~ 1+F*P*Q*lQ*lT + (1+P+Q+lQ+lT|item) + (1+F+P+Q+lQ+lT|subj);

julia> m6 = @btime fit($MixedModel, $m6form, $mrk17, contrasts=$contr);
  8.687 s (249902 allocations: 135.57 MiB)

julia> m6.objective
7147.922051138178

julia> m6.optsum.feval
1092

julia> θ6 = m6.θ;

julia> @benchmark objective(updateL!(setθ!($m6, $θ6)))
BenchmarkTools.Trial: 
  memory estimate:  1.86 KiB
  allocs estimate:  84
  --------------
  minimum time:     7.408 ms (0.00% GC)
  median time:      7.468 ms (0.00% GC)
  mean time:        7.601 ms (0.00% GC)
  maximum time:     23.167 ms (0.00% GC)
  --------------
  samples:          658
  evals/sample:     1
```

### CODE REVIEW

**Comment by dmbates** [comment]
I just realized that this branch is based on the `rankUpdate` branch.  I'm not sure if it is best to squash and merge #331 then rebase and force-push or just to squash and merge #332

**Comment by palday** [comment]
@dmbates  I'll take a look at #331 before this one, sorry for the delay.

File: src/linearmixedmodel.jl [hunk 1]
```diff
@@ -427,8 +431,15 @@ function getθ!(v::AbstractVector{T}, m::LinearMixedModel{T}) where {T}
     if length(v) ≠ length(pmap)
         throw(DimensionMismatch("length(v) = $(length(v)) ≠ length(m.parmap) = $(length(pmap))"))
     end
+    reind = 1
+    λ = first(m.allterms).λ
     for (k, tp) in enumerate(pmap)
-        v[k] = m.allterms[tp[1]].λ[tp[2], tp[3]]
+        tp1 = first(tp)
+        if reind ≠ tp1
+            reind = tp1
```
Comment (palday) [inline src/linearmixedmodel.jl:439]: I don't understand what's going on here -- `reind` gets set but never used.

Comment (palday) [reply src/linearmixedmodel.jl:439]: Oh wait, never mind, it's used for tracking where we are.

**Comment by palday** [comment]
If the CIs pass (i.e. if I haven't messed up the merge), then go for squash and merge :smile: :rocket:

**Comment by palday** [comment]
The next thing I need to do is fix the Cholesky stuff so that Future consistently passes.

**Pull request merged**